packages/ai/ai/src/internal/aiPlan.ts

```diff
@@ -0,0 +1,163 @@
+import * as Cause from "effect/Cause"
+import * as Context from "effect/Context"
+import * as Effect from "effect/Effect"
+import { CommitPrototype } from "effect/Effectable"
+import * as Exit from "effect/Exit"
+import { dual, identity } from "effect/Function"
+import * as Option from "effect/Option"
+import { pipeArguments } from "effect/Pipeable"
+import * as Predicate from "effect/Predicate"
+import * as Schedule from "effect/Schedule"
+import * as Scope from "effect/Scope"
+import type * as AiModel from "../AiModel.js"
+import { AiModels } from "../AiModels.js"
+import type * as AiPlan from "../AiPlan.js"
+
+/** @internal */
+export const TypeId: AiPlan.TypeId = Symbol.for("@effect/ai/AiPlan") as AiPlan.TypeId
+
+/** @internal */
+export const PlanPrototype = {
+  ...CommitPrototype,
+  [TypeId]: TypeId,
+  commit(this: AiPlan.AiPlan<any, any, any>) {
+    return buildPlan(this)
+  },
+  pipe() {
+    return pipeArguments(this, arguments)
+  }
+} as any
+
+const makePlan = <
+  Steps extends ReadonlyArray<AiPlan.AiPlan.Step<any, any, any>>
+>(steps: Steps): AiPlan.AiPlan<any, any, any> => {
+  const self = Object.create(PlanPrototype)
+  self.steps = steps
+  return self
+}
+
+const buildPlan = <Error, Provides, Requires>(
+  plan: AiPlan.AiPlan<Error, Provides, Requires>
+): Effect.Effect<
+  <A, E, R>(effect: Effect.Effect<A, E, R>) => Effect.Effect<A, E, Exclude<R, Provides>>,
+  never,
+  Requires | AiModels
+> =>
+  Effect.map(Effect.context<AiModels | Requires>(), (context) => {
+    const models = Context.get(context, AiModels)
+    return Effect.fnUntraced(function*<A, E, R>(effect: Effect.Effect<A, E, R>) {
+      let exit: Exit.Exit<A, E> | undefined = undefined
+      for (const step of plan.steps) {
+        if (exit !== undefined && Exit.isFailure(exit) && Option.isSome(step.check)) {
+          const check = step.check.value(Cause.squash(exit.cause) as Error)
+          const isFatalError = !(Effect.isEffect(check) ? yield* check : check)
+          if (isFatalError) break
+        }
+        const retryOptions = getRetryOptions(step)
+        exit = yield* Effect.scopedWith((scope) =>
+          models.build(step.model, context).pipe(
+            Scope.extend(scope),
+            Effect.flatMap((context) =>
+              effect.pipe(
+                Option.isSome(retryOptions)
+                  ? Effect.retry(retryOptions.value)
+                  : identity,
+                Effect.provide(context)
+              )
+            ),
+            Effect.exit
```

Comment:
Probably should use Effect.either here, and let defects pass through.

Reply:
Resolved by #4537